File: backend/app/services/property_service.py
from typing import Dict, List, Any, Optional, BinaryIO
import uuid
from datetime import datetime
import logging
import os

# Import the storage service for S3 document handling
from backend.app.services.storage_service import StorageService

# This is a mock implementation for demonstration purposes
# In a real implementation, this would interact with the blockchain

class PropertyService:
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        # Mock database for demonstration
        self.properties = {}
        # Initialize storage service for document handling
        self.storage_service = StorageService()
        
    def register_property(self, address: str, owner_id: str, property_details: Dict[str, Any], documents: List[Dict[str, Any]]) -> str:
        """
        Register a new property on the blockchain.
        
        Args:
            address: Property address
            owner_id: ID of the property owner
            property_details: Additional property details
            documents: List of document dictionaries with keys:
                      - file_obj: File-like object containing document data
                      - filename: Original filename
                      - document_type: Type of document
            
        Returns:
            str: The ID of the newly registered property
        """
        try:
            # Generate a unique property ID
            property_id = f"PROP-{uuid.uuid4().hex[:8].upper()}"
            
            # Upload documents to S3
            document_paths = []
            for doc in documents:
                # Upload document to S3
                s3_path = self.storage_service.upload_document(
                    file_obj=doc["file_obj"],
                    filename=doc["filename"],
                    document_type=doc["document_type"],
                    property_id=property_id
                )
                document_paths.append(s3_path)
            
            # Create property record
            now = datetime.now().isoformat()
            property_record = {
                "id": property_id,
                "address": address,
                "owner_id": owner_id,
                "property_type": property_details.get("type", "unknown"),
                "size": property_details.get("size", 0),
                "rooms": property_details.get("rooms"),
                "description": property_details.get("description"),
                "documents": document_paths,
                "status": "active",
                "created_at": now,
                "updated_at": now
            }
            
            # In a real implementation, this would call the blockchain
            # For now, we'll just store it in our mock database
            self.properties[property_id] = property_record
            
            self.logger.info(f"Property registered: {property_id} with {len(document_paths)} documents")
            return property_id
            
        except Exception as e:
            self.logger.error(f"Error registering property: {str(e)}")
            raise Exception(f"Failed to register property: {str(e)}")
    
    def get_property(self, property_id: str) -> Optional[Dict[str, Any]]:
        """
        Get property details by ID.
        
        Args:
            property_id: The ID of the property to retrieve
            
        Returns:
            Dict or None: Property details if found, None otherwise
        """
        try:
            # In a real implementation, this would query the blockchain
            property_data = self.properties.get(property_id)
            return property_data
        except Exception as e:
            self.logger.error(f"Error retrieving property {property_id}: {str(e)}")
            raise Exception(f"Failed to retrieve property: {str(e)}")
    
    def list_properties(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        """
        List properties with optional filtering.
        
        Args:
            filters: Optional filters to apply
            
        Returns:
            List: List of properties matching the filters
        """
        try:
            # In a real implementation, this would query the blockchain
            if not filters:
                return list(self.properties.values())
            
            # Apply filters
            result = []
            for prop in self.properties.values():
                match = True
                for key, value in filters.items():
                    if key not in prop or prop[key] != value:
                        match = False
                        break
                if match:
                    result.append(prop)
            
            return result
        except Exception as e:
            self.logger.error(f"Error listing properties: {str(e)}")
            raise Exception(f"Failed to list properties: {str(e)}")
    
    def update_property(self, property_id: str, property_details: Dict[str, Any]) -> bool:
        """
        Update property details.
        
        Args:
            property_id: The ID of the property to update
            property_details: Updated property details
            
        Returns:
            bool: True if update was successful, False otherwise
        """
        try:
            # Check if property exists
            if property_id not in self.properties:
                return False
            
            # Update property details
            property_record = self.properties[property_id]
            for key, value in property_details.items():
                if value is not None:
                    property_record[key] = value
            
            # Update timestamp
            property_record["updated_at"] = datetime.now().isoformat()
            
            # In a real implementation, this would update the blockchain
            self.properties[property_id] = property_record
            
            self.logger.info(f"Property updated: {property_id}")
            return True
            
        except Exception as e:
            self.logger.error(f"Error updating property {property_id}: {str(e)}")
            raise Exception(f"Failed to update property: {str(e)}")

File: frontend/src/services/api/propertyService.js
import axios from 'axios';

const API_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000/api';

/**
 * Service for property-related API calls
 */
const propertyService = {
  /**
   * Get all properties with optional filtering
   * @param {Object} filters - Optional filters (owner_id, property_type, etc.)
   * @returns {Promise} - Response with properties list
   */
  getProperties: async (filters = {}) => {
    try {
      // Convert filters object to query string
      const queryParams = new URLSearchParams();
      Object.entries(filters).forEach(([key, value]) => {
        if (value) queryParams.append(key, value);
      });
      
      const queryString = queryParams.toString();
      const url = queryString ? `${API_URL}/properties?${queryString}` : `${API_URL}/properties`;
      
      const response = await axios.get(url);
      return response.data;
    } catch (error) {
      throw error.response?.data || { detail: 'Failed to fetch properties' };
    }
  },

  /**
   * Get property by ID
   * @param {string} propertyId - Property ID
   * @returns {Promise} - Response with property details
   */
  getPropertyById: async (propertyId) => {
    try {
      const response = await axios.get(`${API_URL}/properties/${propertyId}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || { detail: 'Failed to fetch property details' };
    }
  },

  /**
   * Register a new property
   * @param {Object} propertyData - Property data
   * @returns {Promise} - Response with registered property
   */
  registerProperty: async (propertyData) => {
    try {
      const response = await axios.post(`${API_URL}/properties`, propertyData);
      return response.data;
    } catch (error) {
      throw error.response?.data || { detail: 'Failed to register property' };
    }
  },

  /**
   * Update property details
   * @param {string} propertyId - Property ID
   * @param {Object} propertyData - Updated property data
   * @returns {Promise} - Response with updated property
   */
  updateProperty: async (propertyId, propertyData) => {
    try {
      const response = await axios.put(`${API_URL}/properties/${propertyId}`, propertyData);
      return response.data;
    } catch (error) {
      throw error.response?.data || { detail: 'Failed to update property' };
    }
  },

  /**
   * Get property history
   * @param {string} propertyId - Property ID
   * @returns {Promise} - Response with property history
   */
  getPropertyHistory: async (propertyId) => {
    try {
      const response = await axios.get(`${API_URL}/properties/${propertyId}/history`);
      return response.data;
    } catch (error) {
      throw error.response?.data || { detail: 'Failed to fetch property history' };
    }
  },

  /**
   * Verify property ownership
   * @param {string} propertyId - Property ID
   * @param {string} ownerId - Owner ID to verify
   * @returns {Promise} - Response with verification result
   */
  verifyOwnership: async (propertyId, ownerId) => {
    try {
      const response = await axios.get(`${API_URL}/properties/${propertyId}/verify?owner_id=${ownerId}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || { detail: 'Failed to verify ownership' };
    }
  }
};

export default propertyService;